controllers/loadbalancer/nsxtloadbalancer_controller.go

```diff
@@ -0,0 +1,472 @@
+			}
+
+		}()
+	} else if len(ips) == 1 {
+		vip = ips[0]
+	} else {
+		return reconcile.Result{}, fmt.Errorf("got more than 1 VIP for service %s", nsxtLB.Name)
+	}
+
+	// Get the CAPI Machine resources for the cluster.
+	machineList := &clusterv1.MachineList{}
+	if err := ctx.Client.List(
+		ctx, machineList,
+		ctrlclient.InNamespace(ctx.Cluster.Namespace),
+		ctrlclient.MatchingLabels(
+			map[string]string{
+				clusterv1.ClusterLabelName: ctx.Cluster.Name,
+			},
+		)); err != nil {
+		return reconcile.Result{}, errors.Wrapf(
+			err, "failed to get machines for Cluster %s %s/%s for %s",
+			ctx.Cluster.GroupVersionKind(),
+			ctx.Cluster.Namespace,
+			ctx.Cluster.Name,
+			ctx)
+	}
+
+	// Get the control plane machines.
+	controlPlaneMachines := clusterutilv1.GetControlPlaneMachinesFromList(machineList)
+
+	lbMembers := nsxtService.MachinesToLBMembers(controlPlaneMachines)
+	lbPool, err := nsxtService.CreateOrUpdateLBPool(lbName, lbMembers)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var newVirtualServerIDs []string
+	virtualServerID := ""
+	virtualServerExists := false
+	for _, virtualServer := range virtualServers {
+		if virtualServer.DisplayName != nsxtService.GetVirtualServerName(lbName, 6443) {
+			continue
+		}
+
+		virtualServerExists = true
+		virtualServerID = virtualServer.Id
+		break
+	}
+
+	virtualServer := loadbalancer.LbVirtualServer{
+		DisplayName:            nsxtService.GetVirtualServerName(lbName, 6443),
+		Description:            fmt.Sprintf("LoadBalancer VirtualServer managed by CAPV"),
+		IpProtocol:             "TCP",
+		DefaultPoolMemberPorts: []string{strconv.Itoa(6443)},
+		IpAddress:              vip,
+		Ports:                  []string{strconv.Itoa(int(6443))},
+		Enabled:                true,
+		PoolId:                 lbPool.Id,
+	}
+
+	if !virtualServerExists {
+		virtualServer, _, err = nsxtService.Client.ServicesApi.CreateLoadBalancerVirtualServer(nsxtService.Client.Context, virtualServer)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		virtualServer, _, err = nsxtService.Client.ServicesApi.UpdateLoadBalancerVirtualServer(nsxtService.Client.Context, virtualServerID, virtualServer)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+	}
+
+	newVirtualServerIDs = append(newVirtualServerIDs, virtualServer.Id)
+
+	err = nsxtService.AddVirtualServersToLoadBalancer(newVirtualServerIDs, nsxtLB.Spec.LoadBalancerServiceID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	releaseAllocatedVIP = false
+
+	return reconcile.Result{}, nil
+}
+
+// controlPlaneMachineToNSXTLoadBalancer is a handler.ToRequestsFunc to be
+// used to trigger reconcile events for an NSXTLoadBalancer when a CAPI
+// Machine is reconciled and it has IP addresses and is a member of the same
+// control plane that the NSXTLoadBalancer services.
+func (r nsxtlbReconciler) controlPlaneMachineToNSXToadBalancer(o handler.MapObject) []ctrl.Request {
```

Comment:
Hi @yastij,

I bet you can create a common function that can be re-used by this LB and HAProxy that follows a pattern:

```go
ControlPlaneMachineToLoadBalancer(loadBalancerGVK GroupVersionKind) func(handler.MapObject) []ctrl.Request {
  // returns the actual mapping function, using the "loadBalancerGVK" to ensure any loadBalancerRef matches the intended type
}
```